File: frameworks/python-playwright/pages/cart_page.py
from playwright.sync_api import Page, expect

from helpers.steps import Actions, step


class CartPage:
    def __init__(self, page: Page) -> None:
        self.page = page

    def open(self) -> None:
        with step(Actions.CART_OPEN):
            self.page.goto("/cart", wait_until="commit")

    def expect_item(self, product_name: str, quantity: int, unit_price: float) -> None:
        with step(Actions.CART_VERIFY_ITEM):
            items = self.page.get_by_test_id("cart-items")
            expect(items).to_contain_text(product_name)
            expect(items).to_contain_text(str(quantity))
            expect(self.page.get_by_test_id("cart-subtotal")).to_contain_text(
                f"{quantity * unit_price:.2f}"
            )

    def proceed_to_checkout(self) -> None:
        with step(Actions.CART_CHECKOUT):
            self.page.get_by_test_id("cart-checkout").click()

File: frameworks/typescript-playwright/pages/cart_page.ts
import { expect, type Page } from '@playwright/test';
import { ACTIONS, step } from '../helpers/steps.js';

export class CartPage {
  constructor(private readonly page: Page) {}

  async open(): Promise<void> {
    await step(ACTIONS.cartOpen, async () => {
      await this.page.goto('/cart', { waitUntil: 'commit' });
    });
  }

  async expectItem(productName: string, quantity: number, unitPrice: number): Promise<void> {
    await step(ACTIONS.cartVerifyItem, async () => {
      const item = this.page.getByTestId('cart-items').filter({ hasText: productName });
      await expect(item).toBeVisible();
      await expect(item).toContainText(String(quantity));
      await expect(this.page.getByTestId('cart-subtotal')).toContainText(
        (quantity * unitPrice).toFixed(2),
      );
    });
  }

  async proceedToCheckout(): Promise<void> {
    await step(ACTIONS.cartCheckout, async () => {
      await this.page.getByTestId('cart-checkout').click();
    });
  }
}
